src/collect_as.jl

```diff
@@ -1,100 +1,150 @@
-    et = eltype(s)
-    FixedSizeVector{et,V{et}}(s)  # fix underlying storage type
+function collect_as_return_type_helper(::Type{Storage}, ::Type{Vector{E}}) where {Storage <: AbstractVector, E}
+    fsv_type_from_underlying_storage_type(collect_as_storage_type_helper(Storage, E))
+end
+
+function make_fsv_from_tuple(::Type{V}, elems::Tuple) where {V <: DenseVector}
+    stor = collect_as_vector_type_helper(V, elems)
+    ret_type = fsv_type_from_underlying_storage_type(stor)
+    make_abstract_vector_from_tuple(ret_type, elems)
+end
+
+function make_vector_from_tuple(::Type{V}, elems::Tuple) where {V <: DenseVector}
+    stor = collect_as_vector_type_helper(V, elems)
+    ret_type = vector_type_from_underlying_storage_type(stor)
+    make_abstract_vector_from_tuple(ret_type, elems)
+end
+
+function parent_type_with_default(::Type{<:(FixedSizeArray{E, N, T} where {N})}) where {E, T <: DenseVector{E}}
+    T
+end
+function parent_type_with_default(::Type{<:FixedSizeArray{E}}) where {E}
+    default_underlying_storage_type{E}
+end
+function parent_type_with_default(::Type{<:FixedSizeArray})
+    default_underlying_storage_type
+end
+for T ∈ (Vector, optional_memory...)
+    FSA = FixedSizeArray{E, N, T{E}} where {E, N}
+    @eval begin
+        function parent_type_with_default(::Type{$FSA})
+            $T
+        end
+        function parent_type_with_default(::Type{($FSA){E, N} where {E}}) where {N}
+            $T
+        end
+    end
 end
 
-function collect_as_fsa1_from_unknown_length(iterator, ::SpecFSA{1,V}) where {E,V<:DenseVector{E}}
-    E::Type
-    v = collect(E, iterator)::AbstractVector{E}
-    T = FixedSizeVector{E,V}
-    T(v)::T
+function push!!(v::Vector{E}, e::E) where {E}
+    push!(v, e)
+end
+function push!!(v::Vector, e)
+    E = typejoin(typeof(e), eltype(v))
+    ret = Vector{E}(undef, length(v) + 1)
+    ret = copyto!(ret, v)
+    ret[end] = e
+    ret
 end
 
-function collect_as_fsa_impl(iterator, spec::SpecFSA{0}, ::LengthIsKnown)
-    collect_as_fsa0(iterator, spec)::FixedSizeArray{<:Any,0}
+function empty_fsv(::Type{V}, ::Any) where {E, V <: DenseVector{E}}
+    FixedSizeVector{E, V}(undef, 0)
 end
 
-function collect_as_fsa_impl(iterator, spec::SpecFSA, ::LengthIsKnown)
-    collect_as_fsam(iterator, spec)::FixedSizeArray
+function empty_fsv(::Type{V}, iterator) where {V <: DenseVector}
+    let E = eltype(iterator)
+        if isconcretetype(E)
+            FixedSizeVector{E, V{E}}(undef, 0)
+        else
+            let E = Union{}
+                FixedSizeVector{E, V{E}}(undef, 0)
+            end
+        end
+    end
 end
 
-function collect_as_fsa_impl(iterator, spec::SpecFSA{1}, ::LengthIsUnknown)
-    collect_as_fsa1_from_unknown_length(iterator, spec)::FixedSizeVector
+function collect_as_fsv(::Type{V}, iterator) where {V <: DenseVector}
+    es1 = iterate(iterator)  # unroll a bit to avoid unnecessary allocations and help inference
+    T2 = Tuple{Any, Any}
```

Comment:
is this really necessary? This feels like it has to be inferable with a lot less uglyness.

Reply:
OK, now only one iteration is unrolled, which is natural to do, I'd say, given that the behavior for empty iterators is special.